# S10 - Running LLMs Locally
## Exercises

These exercises cover running Large Language Models locally using three popular tools: **Ollama**, **LM Studio**, and **HuggingFace Transformers**. You will learn how to set up each tool, interact with local models programmatically, compare their approaches, and evaluate them on a real NLP task (sentiment analysis).

---

## Exercise 1: Text Generation with Ollama (Easy)

Ollama is a lightweight tool for running LLMs locally. It manages model downloads, quantization, and provides a simple CLI and REST API. Ollama runs models as a local server, exposing an OpenAI-compatible API on `http://localhost:11434`.

Before starting, make sure you have Ollama installed and running on your machine, with a small model pulled (e.g., `gemma3:1b` or `qwen2.5:1.5b`).

**Task:**
1. Use the `requests` library to send a POST request to Ollama's `/api/generate` endpoint
2. Send the prompt: `"Explain what sentiment analysis is in NLP in 2-3 sentences."`
3. Use a small model (e.g., `gemma3:1b`)
4. Set `stream: False` to get the complete response at once
5. Print the model's response text and the total generation time (available in the `total_duration` field of the response, in nanoseconds — convert to seconds)

In [ ]:
import requests
import json

OLLAMA_URL = "http://localhost:11434"

# Send a generation request to Ollama
# Model: "gemma3:1b" (or another small model you have pulled)
# Prompt: "Explain what sentiment analysis is in NLP in 2-3 sentences."
# Print the response and generation time

# Request payload
payload = {
    "model": "gemma3:1b",
    "prompt": "Explain what sentiment analysis is in NLP in 2-3 sentences.",
    "stream": False
}

# Send POST request
response = requests.post(OLLAMA_URL, json=payload)

# Parse JSON response
data = response.json()

# Print model response
print("Model response:\n")
print(data["response"])

# Convert nanoseconds to seconds
total_time_seconds = data["total_duration"] / 1e9

print(f"\nTotal generation time: {total_time_seconds:.3f} seconds")

---

## Exercise 2: Chat Completion with LM Studio (Easy)

LM Studio is a desktop application that lets you download, manage, and run LLMs locally with a graphical interface. Like Ollama, it exposes a local API server — but LM Studio uses the OpenAI-compatible chat completions format at `http://localhost:1234/v1`.

Before starting, make sure you have LM Studio installed, a small model loaded (e.g., a quantized Qwen, Gemma or Phi model), and the local server started.

**Task:**
1. Use the `openai` Python library to connect to LM Studio's local server
2. Set the `base_url` to `http://localhost:1234/v1` and use `api_key="lm-studio"` (any non-empty string works)
3. Send a chat completion request with a system message: `"You are an NLP expert."` and a user message: `"What are the main challenges in sentiment analysis?"`
4. Use the model identifier shown in LM Studio (you can also use `model="local-model"` as a generic identifier)
5. Print the response content and the token usage statistics (prompt tokens, completion tokens, total tokens)

In [2]:
from openai import OpenAI

# Connect to LM Studio's local server
# Send a chat completion request about sentiment analysis challenges
# Print the response and token usage

# Connect to LM Studio local server
client = OpenAI(
    base_url="http://localhost:1234/v1",
    api_key="lm-studio"  # any non-empty string works
)

# Send chat completion request
response = client.chat.completions.create(
    model="local-model",
    messages=[
        {"role": "system", "content": "You are an NLP expert."},
        {"role": "user", "content": "What are the main challenges in sentiment analysis?"}
    ]
)

# Print response content
print("Model response:\n")
print(response.choices[0].message.content)

# Print token usage
print("\nToken usage:")
print(f"Prompt tokens: {response.usage.prompt_tokens}")
print(f"Completion tokens: {response.usage.completion_tokens}")
print(f"Total tokens: {response.usage.total_tokens}")

Model response:

Okay, let's dive into the fascinating and surprisingly complex world of sentiment analysis! As an NLP expert, I can tell you that it’s far more nuanced than simply labeling text as “positive” or “negative.” There are *many* challenges involved. Here’s a breakdown of the main ones, categorized for clarity:

**1. Linguistic Challenges – The Core Problems:**

* **Ambiguity & Context is King:** This is arguably the biggest hurdle. Human language is inherently ambiguous. A word like "sick" can mean genuinely ill or incredibly awesome depending on context. Sentiment analysis systems struggle to grasp this contextual dependency.  
    * **Example:** “That movie was sick!” (positive slang) vs. “I’m feeling really sick.” (negative).
* **Sarcasm & Irony:** These are *devastating* for sentiment analysis. Sarcastic statements deliberately convey the opposite of their literal meaning. Systems often fail to detect this because they don't understand the speaker's intent. 
    * **Exa

---

## Exercise 3: Local Inference with HuggingFace Transformers (Easy)

HuggingFace Transformers is the most widely used Python library for working with pre-trained models. Unlike Ollama and LM Studio (which manage model downloads and serve them via API), HuggingFace loads models directly into your Python process. The `pipeline` API provides a simple high-level interface for common NLP tasks.

**Task:**
1. Use `transformers.pipeline` to create a text generation pipeline with a small model (e.g., `"TinyLlama/TinyLlama-1.1B-Chat-v1.0"` or `"Qwen/Qwen2.5-0.5B-Instruct"`)
2. Send the prompt: `"Sentiment analysis is important because"`
3. Set `max_new_tokens=100` and `do_sample=True` with `temperature=0.7`
4. Print the generated text
5. Measure and print the inference time using `time.time()`

In [3]:
from transformers import pipeline
import time

# Create a text generation pipeline with a small model
# Generate text and measure inference time

# Create text generation pipeline
generator = pipeline(
    "text-generation",
    model="TinyLlama/TinyLlama-1.1B-Chat-v1.0"
)

# Prompt
prompt = "Sentiment analysis is important because"

# Measure inference time
start_time = time.time()

# Generate text
output = generator(
    prompt,
    max_new_tokens=100,
    do_sample=True,
    temperature=0.7
)

end_time = time.time()

# Print generated text
print("Generated text:\n")
print(output[0]["generated_text"])

# Print inference time
print(f"\nInference time: {end_time - start_time:.3f} seconds")

c:\Users\Kike\Programacion\personal\NLP_portfolio_posada_enrique\venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
c:\Users\Kike\Programacion\personal\NLP_portfolio_posada_enrique\venv\Lib\site-packages\huggingface_hub\file_download.py:138: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\Kike\.cache\huggingface\hub\models--TinyLlama--TinyLlama-1.1B-Chat-v1.0. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activat

Generated text:

Sentiment analysis is important because it can help businesses identify customer needs and preferences, and tailor their products and services accordingly. By analyzing the sentiment of customer reviews and feedback, businesses can understand what their customers are saying about their products and services, and make informed decisions about product innovation and marketing strategies.

3. Brand Awareness: By analyzing customer reviews and feedback, businesses can gain insights into their brand's reputation and identify areas where they need to improve.

Inference time: 5.196 seconds


---

## Exercise 4: Comparing the Three Tools (Medium)

Now that you have used all three tools, it is time to compare them systematically. Each tool has different strengths: Ollama is optimized for ease of use via CLI/API, LM Studio provides a GUI with fine-grained control, and HuggingFace offers the deepest integration with the Python ecosystem.

**Task:**
1. Send the **same prompt** to all three tools: `"Classify the sentiment of the following text as Positive, Negative, or Neutral. Only respond with the sentiment label.\n\nText: 'The movie was absolutely wonderful, I loved every minute of it'\n\nSentiment:"`
2. For each tool, record:
   - The response text
   - The approximate response time (use `time.time()` before and after each call)
   - Any qualitative observations (response quality, formatting, etc.)
3. Print a summary table comparing the three tools
4. Write a brief paragraph (as a markdown cell or as a Python comment) explaining which tool you found easiest to use and why

In [5]:
import time
import requests
from openai import OpenAI

results = {}

prompt = """Classify the sentiment of the following text as Positive, Negative, or Neutral. Only respond with the sentiment label.

Text: 'The movie was absolutely wonderful, I loved every minute of it'

Sentiment:"""

# --- Ollama ---

start = time.time()

ollama_response = requests.post(
    "http://localhost:11434/api/generate",
    json={
        "model": "gemma3:1b",
        "prompt": prompt,
        "stream": False
    }
)

end = time.time()

ollama_text = ollama_response.json()["response"].strip()

results["Ollama"] = {
    "response": ollama_text,
    "time": end - start
}

# --- LM Studio ---

client = OpenAI(
    base_url="http://localhost:1234/v1",
    api_key="lm-studio"
)

start = time.time()

lm_response = client.chat.completions.create(
    model="local-model",
    messages=[
        {"role": "system", "content": "You are an NLP expert."},
        {"role": "user", "content": prompt}
    ]
)

end = time.time()

lm_text = lm_response.choices[0].message.content.strip()

results["LM Studio"] = {
    "response": lm_text,
    "time": end - start
}

# --- HuggingFace ---

generator = pipeline(
    "text-generation",
    model="TinyLlama/TinyLlama-1.1B-Chat-v1.0"
)

start = time.time()

hf_output = generator(
    prompt,
    max_new_tokens=20,
    do_sample=True,
    temperature=0.7
)

end = time.time()

hf_text = hf_output[0]["generated_text"].replace(prompt, "").strip()

results["HuggingFace"] = {
    "response": hf_text,
    "time": end - start
}

# Print comparison table

print("\nComparison Results:\n")
print(f"{'Tool':<15}{'Response':<20}{'Time (s)':<10}")
print("-" * 45)

for tool, data in results.items():
    print(f"{tool:<15}{data['response']:<20}{data['time']:.3f}")


Loading weights: 100%|██████████| 201/201 [00:00<00:00, 11750.88it/s]
Both `max_new_tokens` (=20) and `max_length`(=2048) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)



Comparison Results:

Tool           Response            Time (s)  
---------------------------------------------
Ollama         Positive            3.274
LM Studio      Positive            2.118
HuggingFace    Positive

Text: 'I don't really have a favorite color. I just like1.871


**Your comparison:** *(Write your observations here: Which tool did you find easiest to use? Which gave the best response? What are the trade-offs?)*

In my opinion, both LM Studio and Ollama were easier to use than Hugging Face, purely because they need less code to work, however, Hugging Face seems more customizable. If i had to pick the easiest tool between Ollama and LM Studio, I'd pick LM Studio but I do not have another reason apart from the fact that the syntax felt easier for me. Ollama and LM Studio gave the same response, which was correct, while HuggingFace kept generating unrelated text after giving the correct answer. Ollama was the slowest, followed by LM Studio (with more than second of difference), and finally, HugginFace was the fastest one, realtively close with LM Studio.

---

## Exercise 5: Evaluating Local LLMs on Sentiment Analysis (Hard)

In this exercise, you will design and run a proper evaluation of the three tools on a sentiment analysis task. You will use a small labeled dataset, send each sample to each tool, and compute standard classification metrics.

The goal is not just to measure accuracy, but to understand how different local LLM setups handle a real NLP task — including response parsing, consistency, and failure modes.

**Task:**
1. Create a test dataset with at least 10 labeled sentences covering Positive, Negative, and Neutral sentiments (or use a subset from an existing dataset like SST-2 or the financial tweets dataset from the previous lab)
2. Write a function for each tool that takes a text and returns the predicted sentiment label (`"Positive"`, `"Negative"`, or `"Neutral"`)
3. Run all test sentences through each tool
4. For each tool, compute and print:
   - Accuracy
   - Per-class precision, recall, and F1 (use `sklearn.metrics.classification_report`)
   - A confusion matrix
5. Write a final analysis (as markdown or comments) discussing:
   - Which tool performed best on this task and why
   - What failure modes did you observe (e.g., verbose responses, refusal to classify, inconsistent formatting)
   - How you handled parsing the model's response into a clean label
   - Your recommendation for which tool to use for local NLP tasks

In [8]:
import time
import requests
import numpy as np
from openai import OpenAI
from sklearn.metrics import classification_report, confusion_matrix
from transformers import pipeline

# --- Test Dataset ---
test_data = [
    ("I love this movie, it was amazing and fun!", "Positive"),
    ("This was the worst experience I’ve ever had.", "Negative"),
    ("The product is okay, nothing special.", "Neutral"),
    ("Absolutely fantastic performance!", "Positive"),
    ("I am very disappointed with the service.", "Negative"),
    ("It works as expected.", "Neutral"),
    ("The food was delicious and I would come again.", "Positive"),
    ("I hate how slow this app is.", "Negative"),
    ("It is fine, neither good nor bad.", "Neutral"),
    ("What a wonderful surprise, I really enjoyed it!", "Positive"),
]

# --- Prediction Functions ---
def predict_ollama(text):
    """Send text to Ollama and return predicted sentiment label."""
    prompt = f"""Classify sentiment as Positive, Negative, or Neutral.
Only return ONE word.

Text: {text}
Sentiment:"""

    r = requests.post(
        "http://localhost:11434/api/generate",
        json={
            "model": "gemma3:1b",
            "prompt": prompt,
            "stream": False
        }
    )

    output = r.json()["response"].strip().lower()

    if "positive" in output:
        return "Positive"
    elif "negative" in output:
        return "Negative"
    else:
        return "Neutral"

client = OpenAI(
base_url="http://localhost:1234/v1",
api_key="lm-studio"
)

def predict_lmstudio(text):
    """Send text to LM Studio and return predicted sentiment label."""
    prompt = f"""Classify sentiment as Positive, Negative, or Neutral.
Only respond with one word.

Text: {text}
Sentiment:"""
    


    r = client.chat.completions.create(
        model="local-model",
        messages=[
            {"role": "system", "content": "You are a strict sentiment classifier."},
            {"role": "user", "content": prompt}
        ]
    )

    output = r.choices[0].message.content.strip().lower()

    if "positive" in output:
        return "Positive"
    elif "negative" in output:
        return "Negative"
    else:
        return "Neutral"


hf_pipe = pipeline(
    "text-generation",
    model="TinyLlama/TinyLlama-1.1B-Chat-v1.0"
)

def predict_huggingface(text):
    """Send text to HuggingFace pipeline and return predicted sentiment label."""
    prompt = f"""Classify sentiment as Positive, Negative, or Neutral.
Only return one word.

Text: {text}
Sentiment:"""

    out = hf_pipe(
        prompt,
        max_new_tokens=10,
        do_sample=False
    )[0]["generated_text"]

    out = out.lower()

    if "positive" in out:
        return "Positive"
    elif "negative" in out:
        return "Negative"
    else:
        return "Neutral"

# --- Run Evaluation ---
# For each tool, predict sentiment for all test sentences
# Compute and print classification_report and confusion_matrix

y_true = []
y_ollama = []
y_lm = []
y_hf = []

for text, label in test_data:
    y_true.append(label)

    y_ollama.append(predict_ollama(text))
    y_lm.append(predict_lmstudio(text))
    y_hf.append(predict_huggingface(text))

print("\n=== OLLAMA ===")
print(classification_report(y_true, y_ollama))
print(confusion_matrix(y_true, y_ollama))

print("\n=== LM STUDIO ===")
print(classification_report(y_true, y_lm))
print(confusion_matrix(y_true, y_lm))

print("\n=== HUGGINGFACE ===")
print(classification_report(y_true, y_hf))
print(confusion_matrix(y_true, y_hf))

Loading weights: 100%|██████████| 201/201 [00:00<00:00, 12835.80it/s]
The following generation flags are not valid and may be ignored: ['temperature']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
Passing `generation_config` together with generation-related arguments=({'max_new_tokens', 'do_sample'}) is deprecated and will be removed in future versions. Please pass either a `generation_config` object OR all generation parameters explicitly, but not both.
Both `max_new_tokens` (=10) and `max_length`(=2048) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=10) and `max_length`(=2048) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=10) and `m


=== OLLAMA ===
              precision    recall  f1-score   support

    Negative       1.00      1.00      1.00         3
     Neutral       1.00      0.67      0.80         3
    Positive       0.80      1.00      0.89         4

    accuracy                           0.90        10
   macro avg       0.93      0.89      0.90        10
weighted avg       0.92      0.90      0.90        10

[[3 0 0]
 [0 2 1]
 [0 0 4]]

=== LM STUDIO ===
              precision    recall  f1-score   support

    Negative       1.00      1.00      1.00         3
     Neutral       1.00      0.67      0.80         3
    Positive       0.80      1.00      0.89         4

    accuracy                           0.90        10
   macro avg       0.93      0.89      0.90        10
weighted avg       0.92      0.90      0.90        10

[[3 0 0]
 [0 2 1]
 [0 0 4]]

=== HUGGINGFACE ===
              precision    recall  f1-score   support

    Negative       0.00      0.00      0.00         3
     Neutral     

c:\Users\Kike\Programacion\personal\NLP_portfolio_posada_enrique\venv\Lib\site-packages\sklearn\metrics\_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
c:\Users\Kike\Programacion\personal\NLP_portfolio_posada_enrique\venv\Lib\site-packages\sklearn\metrics\_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
c:\Users\Kike\Programacion\personal\NLP_portfolio_posada_enrique\venv\Lib\site-packages\sklearn\metrics\_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parame

**Your evaluation analysis:** *(Discuss your results here: Which tool performed best? What failure modes did you observe? How did you handle response parsing? What is your recommendation?)*

Overall, LM Studio and Ollama performed the best on the sentiment analysis task, both achieving the same metrics, 0.90 accuracy with strong and consistent precision and recall across classes. Their main weakness was occasional confusion between Neutral and Positive cases. In contrast, the HuggingFace (TinyLlama) model performed significantly worse with only 0.40 accuracy, showing a strong bias toward predicting Positive and failing to correctly identify Negative and Neutral sentiments. The most probable reason for this is due to the inconsistent response output (the model kept generating unrelated text in the previous exercise). To handle response parsing, I just looke for the words "Positive, negative or neutral" to map responses to sentiment labels, which worked well for Ollama and LM Studio but was less effective for HuggingFace due to its inconsistent generations. Overall, LM Studio is the most reliable choice for local sentiment classification tasks due to its balance of accuracy, stability, and personal preferences for its syntax.